server/src/test/java/org/elasticsearch/index/engine/ReadOnlyEngineTests.java

```diff
@@ -345,37 +345,54 @@ public boolean keepFullyDeletedSegment(IOSupplier<CodecReader> readerIOSupplier)
                 }
             };
             final EngineConfig config = config(defaultSettings, store, createTempDir(), mergePolicy, null, null, globalCheckpoint::get);
-            final SetOnce<IndexWriter> indexWriter = new SetOnce<>();
-            final IndexWriterFactory indexWriterFactory = (dir, iwc) -> {
-                final IndexWriter writer = new IndexWriter(dir, iwc);
-                indexWriter.set(writer);
-                return writer;
-            };
             String lastSearcherId;
-            try (InternalEngine engine = createEngine(indexWriterFactory, null, null, config)) {
+            Map<String, Engine.Operation> latestOps = new HashMap<>();
+            try (InternalEngine engine = createEngine(null, null, null, config)) {
                 lastSearcherId = ReadOnlyEngine.generateSearcherId(engine.getLastCommittedSegmentInfos());
                 assertNotNull(lastSearcherId);
                 int iterations = randomIntBetween(0, 10);
                 for (int i = 0; i < iterations; i++) {
                     assertThat(ReadOnlyEngine.generateSearcherId(engine.getLastCommittedSegmentInfos()), equalTo(lastSearcherId));
                     final List<Engine.Operation> operations = generateHistoryOnReplica(between(1, 100),
                         engine.getProcessedLocalCheckpoint() + 1L, false, randomBoolean(), randomBoolean());
-                    final long indexWriterSeqNo = indexWriter.get().getMaxCompletedSequenceNumber();
-                    applyOperations(engine, operations);
-                    if (indexWriter.get().getMaxCompletedSequenceNumber() == indexWriterSeqNo) {
-                        assertFalse(
-                            "we don't add documents for no-ops or deletes of non existing documents iff soft-deletes is disabled",
-                            config.getIndexSettings().isSoftDeleteEnabled());
-                        continue;
+                    boolean hasChanges = false;
+                    for (Engine.Operation op : operations) {
+                        if (op instanceof Engine.Index) {
+                            engine.index((Engine.Index) op);
+                            if (latestOps.containsKey(op.id()) == false || latestOps.get(op.id()).seqNo() < op.seqNo()) {
+                                latestOps.put(op.id(), op);
+                                hasChanges = true;
+                            }
+                        } else if (op instanceof Engine.Delete) {
+                            engine.delete((Engine.Delete) op);
+                            final Engine.Operation lastOp = latestOps.get(op.id());
+                            hasChanges |= (lastOp instanceof Engine.Index && lastOp.seqNo() < op.seqNo());
```

Comment:
I struggle to see how this line can have any effect? If we have an `Index` entry in the map, we must have already set `hasChanges = true` above?

Reply:
This line avoids a situation where we have two deletes of the same document id. In that case `hasChanges` should stay intact.